Convert Auto Property To Full Property Refactoring Provider

chborl opened the pull request:
**Customer scenario**

If user has an auto-property, they can convert it to a full-property

**Bugs this fixes:**
new feature

**Risk**

Low.  It's a new feature - doesn't touch existing features.

chborl commented:
This feature is still in development.  This PR is for a preview to get initial feedback.  Things that I still need to add:
- Visual Basic side and tests
- tests for where in the span the refactoring is offered
- additional doc comment tests
- handling directives.  currently, if a directive is around the auto property when expanded it will have the directive around the property only and the field is outside the directive.  need to fix that and add the test.
- additional tests for custom naming rules
- additional tests for using expression body when on single line.
- additional trivia tests to preserve blank lines

Pilchie commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+            }
+
+            // check to see if both property accessors exist and are empty
+            if (!TryGetEmptyAccessors(property, out var emptyGetAccessor, out var emptySetAccessor))
```
We should consider supporting getter-only autoprops too - with a readonly field, etc.

Pilchie commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+            CancellationToken cancellationToken)
+        {
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var namingStyleOptions = options.GetOption(SimplificationOptions.NamingPreferences);
```
Nit - it's a single option, either call it "option" or "preferences"

Pilchie replied:
The method is singular - maybe `namingPreferencesOption`

Pilchie commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var namingStyleOptions = options.GetOption(SimplificationOptions.NamingPreferences);
+            var rules = namingStyleOptions.CreateRules().NamingRules
+                .AddRange(GetDefaultRule(ImmutableArray.Create(new ModifierKind(ModifierKindEnum.IsStatic)), "s_"))
```
Why are we adding these?

chborl replied:
We are getting all of the user-specified naming rules, then adding standard default naming rules for both static and non-static fields.  The standard naming rules are added at the end so they will only be used if the user hasn't specified a preference.  

I added this as a comment to the code.

Pilchie commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+                    isStatic ? DeclarationModifiers.Static : DeclarationModifiers.None, 
+                    Accessibility.Private))
+                {
+                    fieldName = rule.NamingStyle.MakeCompliant(propertyName).Single();
```
Is it safe to `Single` here?  Will there always be exactly one here, not more than one?

CyrusNajmabadi replied:
If MakeCompliant always returns a single string... why is it not typed to return a string?

chborl replied:
@CyrusNajmabadi   MakeCompliant returns IEnumerable<string> and I was using .Single() to convert it to a string.  I have now changed it to .ToString() instead but if there is a better way let me know.

chborl replied:
Upon closer inspection... it is not returning a single string.  Will fix up my code to handle accordingly :)

Pilchie commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+            var workspace = document.Project.Solution.Workspace;
+
+            // add statements to existing accessors
+            var getAccessorStatements = new SyntaxNode[] {
```
Nit: we normally put the braces on new lines (including `}` too).

Pilchie commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+            var setAccessorStatements = new SyntaxNode[] {
+                generator.ExpressionStatement(generator.AssignmentStatement(
+                    generator.IdentifierName(fieldName),
+                    generator.IdentifierName("value"))) };
```
Corner case -in VB it's possible for the setter to define a different name than `value`

Pilchie commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+                .WithAccessorDeclarations(property, new SyntaxNode[] { newGetAccessor, newSetAccessor})
+                .WithAdditionalAnnotations(Formatter.Annotation)
+                .WithAdditionalAnnotations(new SyntaxAnnotation("property"));
+            newProperty = await Formatter.FormatAsync(newProperty, workspace).ConfigureAwait(false);
```
Why do you have to do this here?  Shouldn't it happen automatically when the refactoring is applied at the end?

Pilchie commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+                generator.TypeExpression(propertySymbol.Type), 
+                Accessibility.Private,
+                DeclarationModifiers.From(propertySymbol)).
+                    WithAdditionalAnnotations(Formatter.Annotation);
```
Indentation.

Pilchie replied:
Indentation.

Pilchie commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+            SyntaxGenerator generator, 
+            CancellationToken cancellationToken)
+        {
+            // shell to lang specific to update the accessor.
```
Not sure this comment provides value.  Also, we normally don't call it "shelling out" for overrides in abstract classes.

Pilchie commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+            // shell to lang specific to update the accessor.
+            var newAccessor = UpdateAccessor(accessor, statements);
+
+            // then conver to expression bod
```
missing last letters: "convert", "body"

Pilchie replied:
missing last letters: "convert", "body"

Pilchie commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+            var newAccessor = UpdateAccessor(accessor, statements);
+
+            // then conver to expression bod
+            newAccessor= await ConvertToExpressionBodyIfDesiredAsync(
```
Space before `=`

Pilchie commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,137 @@
+                return accessorDeclarationSyntax.WithSemicolonToken(default);
+            }
+
+            // Should always be able to convert to expression body since we are creating the accessor and know that it only has one line
```
one statement would be more accurate here than one line.

Pilchie commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,137 @@
+            }
+
+            // Should always be able to convert to expression body since we are creating the accessor and know that it only has one line
+            Debug.Assert(accessorDeclarationSyntax.Body.TryConvertToExpressionBody(
```
Debug.Assert will only run in debug builds.  This whole call will be elided in release builds.

Pilchie commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,137 @@
+            return (emptyGetAccessor != null && emptySetAccessor != null);
+        }
+
+        private bool isEmpty(AccessorDeclarationSyntax accessor) 
```
Capitalize method names.

Pilchie commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,137 @@
+        private bool isEmpty(AccessorDeclarationSyntax accessor) 
+            => (accessor.Body == null && accessor.ExpressionBody == null);
+
+        internal override SyntaxNode UpdateAccessor(SyntaxNode accessor, SyntaxNode[] statements)
```
There is never a case when you have multiple statements, right?  Why the array?

chborl commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+                    isStatic ? DeclarationModifiers.Static : DeclarationModifiers.None, 
+                    Accessibility.Private))
+                {
+                    fieldName = rule.NamingStyle.MakeCompliant(propertyName).Single();
```
Yes, MakeCompliant always returns a single string.

chborl commented:
@dotnet/roslyn-ide  Second draft is ready for review.  I added the VB side and additional test cases.  Also updated FieldGenerator.vb to add a specified initializer.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/ConvertAutoPropertyToFullProperty/ConvertAutoPropertyToFullPropertyTests_OptionSets.cs:
```diff
@@ -0,0 +1,116 @@
+﻿using System;
```
Needs copyright.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/ConvertAutoPropertyToFullProperty/ConvertAutoPropertyToFullPropertyTests.cs:
`````diff
@@ -0,0 +1,965 @@
+class goo
+{
+    public int G[||]oo
+````{
`````
````?

CyrusNajmabadi replied:
"````"?

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/ConvertAutoPropertyToFullProperty/ConvertAutoPropertyToFullPropertyTests.cs:
```diff
@@ -0,0 +1,965 @@
+{
+   private int _testgoo;
+
+   public int testf[||]oo {get => _testgoo; set => _testgoo = value; }
```
do not use foo.

CyrusNajmabadi replied:
Do you have a test where we start with "int Prop { get; }" and the user wants "expression bodied properties", so we conver that to "int Prop => _prop"?    If not, can you add it?

chborl replied:
Added this test

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/ConvertAutoPropertyToFullProperty/ConvertAutoPropertyToFullPropertyTests.cs:
```diff
@@ -0,0 +1,965 @@
+            var text = @"
+interface IGoo
+{
+    public int Goo { get; s[||]et; set; }
```
i woudl remove the second set.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/ConvertAutoPropertyToFullProperty/ConvertAutoPropertyToFullPropertyTests_OptionSets.cs:
```diff
@@ -0,0 +1,116 @@
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.NamingStyles;
+using static Microsoft.CodeAnalysis.Diagnostics.Analyzers.NamingStyles.SymbolSpecification;
+using Microsoft.CodeAnalysis.Editing;
```
sort all these.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,191 @@
+        {
+            // C# might have trivia with the accessors that needs to be preserved.  
+            // so we will update the existing accessors instead of creating new ones
+            var accessorListSyntax = ((PropertyDeclarationSyntax)property).AccessorList;
```
can we use generics so that you don't have to do this explicit cast here?

CyrusNajmabadi commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,191 @@
+
+        private (AccessorDeclarationSyntax getAccessor, AccessorDeclarationSyntax setAccessor) GetExistingAccessors(AccessorListSyntax accessorListSyntax)
+        {
+            AccessorDeclarationSyntax getter = null;
```
could all be simplified to: ```return (accessorListSyntax.Accessors.FirstOrDefault(a => a.IsKind(GetAccessorDeclaration), ...)```

CyrusNajmabadi commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,191 @@
+            return (getAccessor: getter, setAccessor: setter);
+        }
+
+        private static void GetExistingAccessors(AccessorListSyntax accessorListSyntax, ref AccessorDeclarationSyntax getAccessor, ref AccessorDeclarationSyntax setAccessor)
```
Why do you need this?

CyrusNajmabadi commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,191 @@
+        }
+
+        private async Task<SyntaxNode> AddStatementsToAccessorAsync(
+                Document document,
```
dedent.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,191 @@
+                newAccessor,
+                cancellationToken).ConfigureAwait(false);
+
+            return await Formatter.FormatAsync(newAccessor, document.Project.Solution.Workspace).ConfigureAwait(false);
```
i would not format here.  Formatting hsould happen automatically.  If that's not happening, you should just add a Formatter.Annotation to these nodes before returning them.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,191 @@
+
+            // Should always be able to convert to expression body since we are creating the accessor 
+            // and know that it only has one statement
+            Debug.Assert(ableToConvert);
```
This it not true. For example, you might be in C# 5.  We won't be able to conver to an expression body there.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,191 @@
+            => accessor.Body == null && accessor.ExpressionBody == null;
+
+        internal override string GetUniqueName(string fieldName, IPropertySymbol property)
+            => NameGenerator.GenerateUniqueName(fieldName, n => !property.ContainingType.GetMembers(n).Any());
```
This is not C# specific, can this move up to the base type?

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,220 @@
+                        document,
+                        property,
+                        root,
+                        cancellationToken)));
```
this last bit of wrapping seems unnecessary.

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,220 @@
+
+            var propertySymbol = semanticModel.GetDeclaredSymbol(property) as IPropertySymbol;
+
+            var members = propertySymbol.ContainingType.GetMembers()
```
no need to wrap here.

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,220 @@
+            var propertySymbol = semanticModel.GetDeclaredSymbol(property) as IPropertySymbol;
+
+            var members = propertySymbol.ContainingType.GetMembers()
+                .Where(n => n.Kind == SymbolKind.Field);
```
.GetMembers().OfType<IFieldSymbol>();

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,220 @@
+            var members = propertySymbol.ContainingType.GetMembers()
+                .Where(n => n.Kind == SymbolKind.Field);
+
+            IFieldSymbol fieldSymbol;
```
```var field = fields.FirstOrDefault(f=> propertySymbol.Equals(f.AssociatedProperty));```

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,220 @@
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken)
+                .ConfigureAwait(false);
+
+            var propertySymbol = semanticModel.GetDeclaredSymbol(property) as IPropertySymbol;
```
you did an 'as', but did not check for null.

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,220 @@
+            CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var propertySymbol = semanticModel.GetDeclaredSymbol(property) as IPropertySymbol;
```
Get the property symbol once at the top level and pass to all your methods that use it.

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,220 @@
+            SyntaxNode root,
+            CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
```
get semantic model once at top level, then pass to all your helpers that need it.

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,220 @@
+            foreach (var rule in rules)
+            {
+                if (rule.SymbolSpecification.AppliesTo(
+                    new SymbolKindOrTypeKind(SymbolKind.Field), 
```
Fix indentation.

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,220 @@
+                fieldName,
+                generator,
+                cancellationToken)
+                    .ConfigureAwait(false);
```
please put .ConfitureAwait on the previous line.  (also, consider just making entire thing a single line).

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,220 @@
+            var containingType = GetTypeBlock(propertySymbol.ContainingType.DeclaringSyntaxReferences.FirstOrDefault().GetSyntax(cancellationToken));
+
+            editor.ReplaceNode(containingType, (currentTypeDecl, _) =>
+                {
```
no need for the body or 'return', this is short enough to just have a simple lambda.

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,220 @@
+        {
+            public ConvertAutoPropertyToFullPropertyCodeAction(
+                string Title, 
+                Func<CancellationToken, 
```
remove wrapping.

chborl replied:
Missed this, but fixed in next round

rchande replied:
?

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,220 @@
+        private class ConvertAutoPropertyToFullPropertyCodeAction : CodeAction.DocumentChangeAction
+        {
+            public ConvertAutoPropertyToFullPropertyCodeAction(
+                string Title, 
```
"title"

chborl replied:
Missed this, but fixed in next round

rchande replied:
Doesn't look fixed?

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/ConvertAutoPropertyToFullProperty/VisualBasicConvertAutoPropertyToFullProperty.vb:
```diff
@@ -0,0 +1,104 @@
+
+        Friend Overrides Function GetProperty(token As SyntaxToken) As SyntaxNode
+            Dim containingProperty = token.Parent.FirstAncestorOrSelf(Of PropertyStatementSyntax)
+            If (containingProperty Is Nothing) Then
```
no need for parens.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/ConvertAutoPropertyToFullProperty/VisualBasicConvertAutoPropertyToFullProperty.vb:
```diff
@@ -0,0 +1,104 @@
+
+            ' Offer this refactoring anywhere in the signature of the property.
+            Dim position = token.SpanStart
+            If (position < start) Then
```
no need for parens.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/ConvertAutoPropertyToFullProperty/VisualBasicConvertAutoPropertyToFullProperty.vb:
```diff
@@ -0,0 +1,104 @@
+            Return False
+        End Function
+
+        Friend Overrides Function GetUniqueName(fieldName As String, propertySymbol As IPropertySymbol) As String
```
For VB do we even want to generate a unique name?  shouldn't we always just use the actual VB name that was created.  There may be existing code that refers to that name, so we don't want to break that code.

chborl replied:
Changed the method to return the name of the implicit backing field.

CyrusNajmabadi commented on src/Workspaces/VisualBasic/Portable/CodeGeneration/FieldGenerator.vb:
```diff
@@ -78,7 +78,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeGeneration
-            Dim initializer = GenerateEqualsValue(field)
+            Dim initializerNode = TryCast(CodeGenerationFieldInfo.GetInitializer(field), ExpressionSyntax)
+            Dim initializer As EqualsValueSyntax
+            If (initializerNode IsNot Nothing) Then
```
no parens needed.

CyrusNajmabadi commented on src/Workspaces/VisualBasic/Portable/CodeGeneration/FieldGenerator.vb:
```diff
@@ -78,7 +78,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeGeneration
+                initializer = SyntaxFactory.EqualsValue(initializerNode)
+            Else
+                initializer = GenerateEqualsValue(field)
+            End If
```
consider just writing this as:  ```dim initializer = if (initializeNode is nothing, ..., ...)```

CyrusNajmabadi commented:
Are there tests that validate that the cursor is on the right span?  i.e. that we don't offer this if you're before the first modifier, or on the attributes?

chborl commented:
@dotnet/roslyn-ide

rchande commented on src/EditorFeatures/CSharpTest/ConvertAutoPropertyToFullProperty/ConvertAutoPropertyToFullPropertyTests.cs:
```diff
@@ -0,0 +1,985 @@
+class goo
+{
+    // Comments before
+    public int G[||]oo { get; private set; } //Comments during
```
Worth testing with trivia on the `get` or `set`?

chborl replied:
I did some ad hoc testing on this and it works fine with trivia before and after both get and set.

rchande commented on src/EditorFeatures/CSharpTest/ConvertAutoPropertyToFullProperty/ConvertAutoPropertyToFullPropertyTests.cs:
```diff
@@ -0,0 +1,985 @@
+{
+    private int _goo;
+
+    public int Goo { get => _goo; set => _goo = value; }
```
Is there any way this could generate 
```
Goo
{
   get => ...
   set => ...
}
```

rchande replied:
(Don't like the whole thing on one line)

rchande commented on src/EditorFeatures/CSharpTest/ConvertAutoPropertyToFullProperty/ConvertAutoPropertyToFullPropertyTests.cs:
```diff
@@ -0,0 +1,985 @@
+
+        [Fact, Trait(Traits.Feature, Traits.Features.ConvertAutoPropertyToFullProperty)]
+        public async Task WithExpressionBodyWhenOnSingleLine()
+        {
```
Is this input supposed to be the same as the previous test? Should there be a test for PreferExpressioNBodiedAccessorsWhenOnSingleLIne with the original Goo definition spanning multiple lines?

chborl replied:
Added this test

dpoeschl commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,199 @@
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(context.Span.Start);
```
It'd be cool if we used the whole span and applied it to everything touched by the span. But this is fine until someone asks for it. :smile:

CyrusNajmabadi replied:
See one of my top coments.  This will find the first token of a property if you're in the doc comment for it.   We should pass along the user position to GetProperty and ensure that it's really within the header region of the property.

dpoeschl commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,199 @@
+        internal bool IsValidAutoProperty(SyntaxNode property, IPropertySymbol propertySymbol)
+        {
+            var fields = propertySymbol.ContainingType.GetMembers().OfType<IFieldSymbol>();
+            var field = fields.FirstOrDefault(f => propertySymbol.Equals(f.AssociatedSymbol));
```
Don't have to change it, but `return fields.Any(f => ...)` would avoid the subsequent null check.

dpoeschl commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,199 @@
+                    DiagnosticSeverity.Hidden));
+        }
+
+        private string GenerateFieldName(IPropertySymbol property, ImmutableArray<NamingRule> rules)
```
Should probably only be on the C# side, and then you won't make an unnecessary one for VB that just gets ignored later.

dpoeschl commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,199 @@
+        /// for both static and non-static fields.  The standard naming rules are added at the end 
+        /// so they will only be used if the user hasn't specified a preference.
+        /// </summary>
+        private static async Task<ImmutableArray<NamingRule>> GetNamingRulesAsync(
```
actually, all the NamingRules stuff should move to the C# side

chborl commented:
@dpoeschl , I moved all of the C# naming rules methods to the CSharp file to simplify getting the field name on the VB side.

dpoeschl commented:
retest perf_correctness_prtest please

chborl commented:
@dpoeschl , FYI - I cleaned up the code a bit more.

dpoeschl commented:
@Pilchie I think this is ready for Ask Mode.

dpoeschl commented:
retest windows_debug_vs-integration_prtest please

Pilchie commented:
Do we know what the deal is with the CLA?  I'm willing to mark approved once that is sorted out.

dpoeschl commented:
retest windows_release_vs-integration_prtest please

dpoeschl commented:
She had already signed it (see the tagging in https://github.com/dotnet/roslyn/pull/18718) but then she changed her GH handle which we guess broke the system.

Pilchie commented:
Can't be signed again?

dnfclas commented:
__@chborl__, thanks for signing the contribution license agreement. We will now validate the agreement and then the pull request.
<br />_Thanks, .NET Foundation Pull Request Bot_

dpoeschl commented:
@Pilchie CLA problem resolved.

CyrusNajmabadi commented on src/EditorFeatures/TestUtilities/Traits.cs:
```diff
@@ -123,6 +123,7 @@ public static class Features
             public const string CodeModelMethodXml = "CodeModel.MethodXml";
             public const string CommentSelection = nameof(CommentSelection);
             public const string Completion = nameof(Completion);
+            public const string ConvertAutoPropertyToFullProperty = nameof(ConvertAutoPropertyToFullProperty);
```
This is a code action right?  Should probably go up in the code action section.  (but i don't care that much :))

CyrusNajmabadi commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,226 @@
+                return null;
+            }
+
+            var start = containingProperty.AttributeLists.Count > 0
```
something to consider for the future.  i imagine we've duplicated this chuck of a code in a few features now.  we probably need to just extract it into a service/extension that answers "is this in the header portion of the declaration?"

CyrusNajmabadi commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,226 @@
+
+            // Offer this refactoring anywhere in the signature of the property
+            var position = token.SpanStart;
+            if (position < start || position > containingProperty.Identifier.Span.End)
```
this doesn't feel quite right.  Have we ensured that hte user's original position is within the token that we're looking at?  otherwise, the user's positino could be in the preceding trivia and we woudl count it as within the header here.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,226 @@
+            CancellationToken cancellationToken)
+        {
+            const string defaultStaticFieldPrefix = "s_";
+            const string defaultFieldPrefix = "_";
```
i'm not sure we shoudl be making that choice for users.  Roslyn itself had no prefix on fields for tons of versions.  So, IMO, if there is no naming rule setup, then all we should do is lowercase the property name.  If that conflicts *then* add the _.  If tehre is a naming rule, we should respect it.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,226 @@
+                    property.IsStatic ? DeclarationModifiers.Static : DeclarationModifiers.None,
+                    Accessibility.Private))
+                {
+                    fieldName = rule.NamingStyle.MakeCompliant(propertyName).First();
```
nit: @dpoeschl it seems unfortunate to have to allocate all these iterators because MakeCompliant sometimes wants to return two values.  Consider having MakeCompliant return a tuple in the future.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,226 @@
+        }
+
+        internal override (SyntaxNode newGetAccessor, SyntaxNode newSetAccessor) GetNewAccessors(
+            DocumentOptionSet options, SyntaxNode property, 
```
Consider making the based typed parameterized on things like TPropertyDeclaration.  That way you don't have to do the casts below.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,226 @@
+            }
+
+            if (!accessorDeclarationSyntax.Body.TryConvertToExpressionBody(
+                accessorDeclarationSyntax.Kind(), accessor.SyntaxTree.Options, preference,
```
indent these further.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,226 @@
+        internal SyntaxNode AddStatement(SyntaxNode accessor, SyntaxNode statement)
+        {
+            var blockSyntax = SyntaxFactory.Block(
+                SyntaxFactory.Token(SyntaxKind.OpenBraceToken).WithLeadingTrivia(SyntaxFactory.CarriageReturnLineFeed),
```
document that the explicit CRLF is so that the block isn't just kept on a single line by the formatter

CyrusNajmabadi commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,226 @@
+                .WithAdditionalAnnotations(Formatter.Annotation);
+        }
+
+        internal SyntaxNode AddStatement(SyntaxNode accessor, SyntaxNode statement)
```
Accessor shoudl always be strongly typed as AccessorDeclarationSyntax.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,226 @@
+        }
+
+        internal override SyntaxNode ConvertPropertyToExpressionBodyIfDesired(
+            DocumentOptionSet options, SyntaxNode property)
```
see above comments about making this generic.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,226 @@
+
+        internal ExpressionBodyPreference GetPropertyExpressionBodyPreference(DocumentOptionSet options)
+            => options.GetOption(CSharpCodeStyleOptions.PreferExpressionBodiedProperties).Value;
+
```
extra newline.

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,117 @@
+                    GetPropertyWithoutInitializer(property),
+                    accessorTuple.newSetAccessor == null
+                        ? new SyntaxNode[] { accessorTuple.newGetAccessor }
+                        : new SyntaxNode[] { accessorTuple.newGetAccessor, accessorTuple.newSetAccessor })
```
i think it *might* be possible to just use "new [] { }" instead of "new SyntaxNode[] { }"

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,117 @@
+                        : new SyntaxNode[] { accessorTuple.newGetAccessor, accessorTuple.newSetAccessor })
+                .WithLeadingTrivia(property.GetLeadingTrivia());
+            fullProperty = ConvertPropertyToExpressionBodyIfDesired(options, fullProperty);
+            var editor = new SyntaxEditor(root, workspace);
```
nit.  i think a newline above this line might be nice.

CyrusNajmabadi replied:
nit: consider extracting out all the code to produce the fullProperty into its own method.

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,117 @@
+                .WithLeadingTrivia(property.GetLeadingTrivia());
+            fullProperty = ConvertPropertyToExpressionBodyIfDesired(options, fullProperty);
+            var editor = new SyntaxEditor(root, workspace);
+            editor.ReplaceNode(property, fullProperty.WithAdditionalAnnotations(Formatter.Annotation));
```
earlier i saw code that attached the formatting annotation to a sub node.  that should not be unnecessary if you're unilaterally formatting the full property.

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,117 @@
+                propertySymbol.ContainingType.DeclaringSyntaxReferences.FirstOrDefault().GetSyntax(cancellationToken));
+            editor.ReplaceNode(containingType, (currentTypeDecl, _)
+                => CodeGenerator.AddFieldDeclaration(currentTypeDecl, newField, workspace)
+                .WithAdditionalAnnotations(Formatter.Annotation));
```
this is very heavyweight.  you're essentialy formatting the *entire* type declaration.  is that necessary?

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,117 @@
+
+            context.RegisterRefactoring(
+                new ConvertAutoPropertyToFullPropertyCodeAction(
+                    FeaturesResources.Convert_to_full_property,
```
title doesn't need to be supplied here.  it can just be inlined in the constructor for ConvertAutoPropertyToFullPropertyCodeAction

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/ConvertAutoPropertyToFullProperty/VisualBasicConvertAutoPropertyToFullProperty.vb:
```diff
@@ -0,0 +1,105 @@
+    Friend Class VisualBasicConvertAutoPropertyToFullPropertyCodeRefactoringProvider
+        Inherits AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider
+
+        Private Const Underscore As String = "_"
```
hrmm... not sure what this is needed for.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/ConvertAutoPropertyToFullProperty/VisualBasicConvertAutoPropertyToFullProperty.vb:
```diff
@@ -0,0 +1,105 @@
+                containingProperty.SpanStart)
+
+            ' Offer this refactoring anywhere in the signature of the property.
+            Dim position = token.SpanStart
```
as before, we should take in hte actual user position.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/ConvertAutoPropertyToFullProperty/VisualBasicConvertAutoPropertyToFullProperty.vb:
```diff
@@ -0,0 +1,105 @@
+        ''' any existing references to this field.
+        ''' </summary>
+        Friend Overrides Function GetFieldNameAsync(document As Document, propertySymbol As IPropertySymbol, cancellationToken As CancellationToken) As Task(Of String)
+            Return Task.FromResult(Underscore + propertySymbol.Name)
```
Two things.  You can just use "_" here instead of having the constant.  Constnats are good when the name tells you something interesting about hte value.  like "Second_in_a_year = 31_536_000".  They're also good if you're going to use the value several times.  When you only use it once, and the name of the variable is exactly the same as the value, they're not needed.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/ConvertAutoPropertyToFullProperty/VisualBasicConvertAutoPropertyToFullProperty.vb:
```diff
@@ -0,0 +1,105 @@
+            Dim propertySyntax = DirectCast(propertyNode, PropertyStatementSyntax)
+
+            Dim setAccessor As SyntaxNode
+            If IsReadOnly(propertySyntax) Then
```
```if not propertySyntax.Modifiers.Any(SyntaxKind.ReadOnlyKeyword)```

CyrusNajmabadi commented:
Ah, you merged during my review.  That's ok :)  I would try to clean up a few of these comments.  But overall it looks quite good. Nice work Cheryl!
